### PR TITLE
fix: Wait until istio cni files appear before copying them

### PR DESCRIPTION
**Pull request by aerosouund**
**What this PR does / why we need it**:

This fixes the issue that appeared with the rewrite of the istio cluster option to go: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/12677/pull-kubevirt-e2e-k8s-1.31-sig-network/1829308375628779520
A brief description of the issue is that during the rewrite the code that copied the istio cni files to /etc/cni/net.d was removed because both issues referenced were closed and because it caused cluster-up to crash

```bash
function copy_istio_cni_conf_files() {
    if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ] && [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
        for nodeNum in $(seq -f "%02g" 1 $KUBEVIRT_NUM_NODES); do
            $ssh node${nodeNum} -- "until ls /etc/cni/multus > /dev/null 2>&1; do sleep 1; done"
            $ssh node${nodeNum} -- sudo cp -uv /etc/cni/multus/net.d/*istio*.conf /etc/cni/net.d/
        done
    fi
}
```

The reason why it would crash is because this code was previously being called after a period of waiting to allow for istio cni files to appear (not deliberately, the code was waiting for something else)

```bash
    until wait_for_cnao_ready && wait_for_istio_ready && wait_for_cdi_ready && wait_for_multus_ready && wait_for_aaq_ready && wait_for_kwok_ready; do
        echo "Waiting for cluster components..."
        sleep 5
    done

    # FIXME: remove 'copy_istio_cni_conf_files()' as soon as [1] and [2] are resolved
    # [1] https://github.com/kubevirt/kubevirtci/issues/906
    # [2] https://github.com/k8snetworkplumbingwg/multus-cni/issues/982
    copy_istio_cni_conf_files
```

this pull request introduces this wait again after ensuring that the cni files appeared on the hosts

It also updates CNAO manifests deployed by the gocli to v0.94.1 which is the version used before using the gocli in deploying those manifests.

As well as remove this part from the yaml manifest of istio operator with CNAO

```yaml
      cniConfFileName: "istio-cni.conf"
    sidecarInjectorWebhook:
      injectedAnnotations:
        "k8s.v1.cni.cncf.io/networks": istio-cni
```
This part was not previously in the manifest and was incorrectly added in #1217 
The effect of this is an error during VMI sync from passt being unable to bind to port 22 and the VMI being stuck in `Scheduled` state

```
passt --one-off --socket /var/run/libvirt/qemu/run/passt/1-kubevirt-test-defaul-ua-default.socket --pid /var/run/libvirt/qemu/run/passt/1-kubevirt-test-defaul-ua-default-passt.pid --interface eth0 --log-file /var/run/kubevirt/passt.log --tcp-ports '~15000,~15001,~15004,~15006,~15008,~15009,~15020,~15021,~15053,~15090,1500,1501,22') unexpected exit status 1: Failed to bind port 22 (Permission denied) for option '-t ~15000,~15001,~15004,~15006,~15008,~15009,~15020,~15021,~15053,~15090,1500,1501,22', exiting\n'
```

And lastly this PR also omits multus from the CNAO CR being deployed in case multus is requested in the cluster to be installed on its own through the `KUBEVIRT_WITH_MULTUS_V3` variable, by moving this logic

```bash
function create_network_addons_config() {
    local nac="/opt/cnao/network-addons-config-example.cr.yaml"
    if [ "$KUBEVIRT_WITH_MULTUS_V3" == "true" ]; then
        local no_multus_nac="/opt/cnao/no-multus-nac.yaml"
        $ssh node01 -- "awk '!/multus/' ${nac} | sudo tee ${no_multus_nac}"
        nac=${no_multus_nac}
    fi

     $kubectl apply -f ${nac}
}
```

into this go logic

```golang
      if path == "manifests/network-addons-config-example.cr.yaml" && o.multusEnabled {
	      re := regexp.MustCompile("(?m)[\r\n]+^.*multus.*$")
	      res := re.ReplaceAllString(string(yamlDoc), "")
	      yamlDoc = []byte(res)
      }
      
      obj, err := k8s.SerializeIntoObject(yamlDoc)
      if err != nil {
	      logrus.Info(err.Error())
	      continue
      }
      if err := o.client.Apply(obj); err != nil {
	      return fmt.Errorf("error applying manifest %s", err)
      }
      }
```

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by aerosouund** [comment]
cc: @brianmcarey

File: cluster-up/cluster/k8s-provider-common.sh [hunk 1]
```diff
@@ -14,7 +14,16 @@ function deploy_kwok() {
     fi
 }
 
-
+# copy_istio_cni_conf_files copy the generated Istio CNI net conf file
+# (at '/etc/cni/multus/net.d/') to where Multus expect CNI net conf files ('/etc/cni/net.d/')
+function copy_istio_cni_conf_files() {
+    if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ] && [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
+        for nodeNum in $(seq -f "%02g" 1 $KUBEVIRT_NUM_NODES); do
+            $ssh node${nodeNum} -- "until ls /etc/cni/multus > /dev/null 2>&1; do sleep 1; done"
```
Comment (brianmcarey) [inline cluster-up/cluster/k8s-provider-common.sh:22]: Is there a danger here that it could sleep forever?

Comment (aerosouund) [reply cluster-up/cluster/k8s-provider-common.sh:22]: its unlikely, since if istio is deployed with cnao the cni file then it will create this cni file. if it couldn't then it would probably crash before getting here

but i will add a timeout for extra safety

**Review by brianmcarey — COMMENTED** [review]
/approve

@aerosouund you mentioned that you ran the sig-network tests locally against this and the tests looked ok?

**Comment by aerosouund** [comment]
> /approve
> 
> 
> 
> @aerosouund you mentioned that you ran the sig-network tests locally against this and the tests looked ok?

Yeah indeed, to be exact not all. The ones that were failing. I.E the ones in the file vmi_istio.go

File: cluster-provision/gocli/opts/cnao/cnao.go [hunk 2]
```diff
@@ -43,6 +46,12 @@ func (o *cnaoOpt) Exec() error {
 					continue
 				}
 
+				if path == "manifests/network-addons-config-example.cr.yaml" && o.multusEnabled {
```
Comment (xpivarc) [inline cluster-provision/gocli/opts/cnao/cnao.go:49]: Can we have a test for this?

**Comment by dhiller** [comment]
@EdDev all we are asking here is to verify whether the changes to the manifests are reasonable - we will make sure that the other changes do work.

**Comment by dhiller** [comment]
Two commits are to look at here:
* https://github.com/kubevirt/kubevirtci/pull/1262/commits/10f7aaf67888d50e1db3a48e67648ab56464e0c0
* https://github.com/kubevirt/kubevirtci/pull/1262/commits/7b8b9e8d9a3566a5157d742376627d55beee4255

File: cluster-up/cluster/k8s-provider-common.sh [hunk 3]
```diff
@@ -95,4 +104,6 @@ function up() {
         echo "Waiting for cluster components..."
         sleep 5
     done
+
+    copy_istio_cni_conf_files
```
Comment (ormergi) [inline cluster-up/cluster/k8s-provider-common.sh:108]: Please add the previous comment and links to tracking issue.

File: cluster-provision/gocli/cmd/run.go [hunk 4]
```diff
@@ -812,7 +812,7 @@ func provisionK8sOptions(sshClient libssh.Client, k8sClient k8s.K8sDynamicClient
 	}
 
 	if n.CNAO {
-		cnaoOpt := cnao.NewCnaoOpt(k8sClient, sshClient)
+		cnaoOpt := cnao.NewCnaoOpt(k8sClient, sshClient, n.Multus)
```
Comment (ormergi) [inline cluster-provision/gocli/cmd/run.go:815]: CNAO should also deploy the dynamic-network-controller for kubevirt/kubevirt e2e tests. 
Please ensure the current behavior remains.

File: cluster-provision/gocli/opts/cnao/manifests/network-addons-config-example.cr.yaml [hunk 5]
```diff
@@ -9,3 +9,5 @@ spec:
   kubeSecondaryDNS: {}
   linuxBridge: {}
   macvtap: {}
+  multus: {}
```
Comment (ormergi) [inline cluster-provision/gocli/opts/cnao/manifests/network-addons-config-example.cr.yaml:12]: This change belong to previous commit.

File: cluster-up/cluster/k8s-provider-common.sh [hunk 6]
```diff
@@ -19,7 +19,7 @@ function deploy_kwok() {
 function copy_istio_cni_conf_files() {
     if [ "$KUBEVIRT_DEPLOY_ISTIO" == "true" ] && [ "$KUBEVIRT_WITH_CNAO" == "true" ]; then
         for nodeNum in $(seq -f "%02g" 1 $KUBEVIRT_NUM_NODES); do
-            $ssh node${nodeNum} -- "until ls /etc/cni/multus > /dev/null 2>&1; do sleep 1; done"
+            timeout 60 $ssh node${nodeNum} -- "until ls /etc/cni/multus > /dev/null 2>&1; do sleep 1; done"
```
Comment (ormergi) [inline cluster-up/cluster/k8s-provider-common.sh:22]: This change is not related to this PR according to title and description.

Please move this change to new PR.

**Review by ormergi — COMMENTED** [review]
Hi, thanks of the PR, please see my comment below and inline comments.

In general it seem to do more then what the title and description say, such as CNAO downgrade, and adding timeout for changes around istio files.
For next time, please make sure a PR does exactly what it describe, and if necessary break it down to smaller PRs.

Regarding commit "fix: Use CNAO v0.94", such changes as version upgrade/downgrade for a component should be done in a self contained PR.
Having such changes makes code-review easier and keep reviewers from being distracted by other changes (noise).
In addition it enable quick revert in case of an issue.
For next time, post separate PR for CNAO version change.

File: cluster-up/cluster/k8s-provider-common.sh (hunk 1 above)
Comment (ormergi) [reply cluster-up/cluster/k8s-provider-common.sh]: Please do this change in separate PR.

File: cluster-up/cluster/k8s-provider-common.sh (hunk 3 above)
Comment (ormergi) [reply cluster-up/cluster/k8s-provider-common.sh:108]: FWIU changes should be exactly as before, initially introduced by https://github.com/kubevirt/kubevirtci/pull/932

File: cluster-up/cluster/k8s-provider-common.sh (hunk 6 above)
Comment (aerosouund) [reply cluster-up/cluster/k8s-provider-common.sh:22]: Indeed, but it was added as a part of a previous review: https://github.com/kubevirt/kubevirtci/pull/1262#discussion_r1740808214

Comment (ormergi) [reply cluster-up/cluster/k8s-provider-common.sh:22]: Nevertheless, this change is not related to this PR, right?

Comment (aerosouund) [reply cluster-up/cluster/k8s-provider-common.sh:22]: Yes its not, maybe we can separate it another PR indeed

File: cluster-up/cluster/k8s-provider-common.sh (hunk 3 above)
Comment (aerosouund) [reply cluster-up/cluster/k8s-provider-common.sh:111]: added them again

File: cluster-provision/gocli/opts/cnao/manifests/network-addons-config-example.cr.yaml (hunk 5 above)
Comment (aerosouund) [reply cluster-provision/gocli/opts/cnao/manifests/network-addons-config-example.cr.yaml:12]: if you are referring to the multus issue then it was squashed

File: cluster-provision/gocli/cmd/run.go (hunk 4 above)
Comment (aerosouund) [reply cluster-provision/gocli/cmd/run.go:815]: this parameter is controlled by the overall `KUBEVIRT_WITH_MULTUS_V3` param
from a callers perspective everything will work the same.
and what it previously was is that if this variable is passed multus is deployed separately. not by CNAO.
see this comment i left for or shoval inquiring about the same thing: https://github.com/kubevirt/kubevirtci/commit/7b8b9e8d9a3566a5157d742376627d55beee4255#r146222431

of if you're referring to something else let me know

File: cluster-up/cluster/k8s-provider-common.sh [hunk 7]
```diff
@@ -14,7 +14,16 @@ function deploy_kwok() {
     fi
 }
 
-
+# copy_istio_cni_conf_files copy the generated Istio CNI net conf file
+# (at '/etc/cni/multus/net.d/') to where Multus expect CNI net conf files ('/etc/cni/net.d/')
+function copy_istio_cni_conf_files() {
```
Comment (xpivarc) [inline cluster-up/cluster/k8s-provider-common.sh:19]: @aerosouund Is this something new or did we have this before?

File: cluster-provision/gocli/cmd/run.go (hunk 4 above)
Comment (xpivarc) [inline cluster-provision/gocli/cmd/run.go:815]: @aerosouund Can you please explain the reasoning to the code as well. This was something that took me time to connect. Please enhance the commit message as well

File: cluster-provision/gocli/opts/istio/manifests/istio-operator-with-cnao.cr.yaml [hunk 8]
```diff
@@ -22,7 +22,3 @@ spec:
        - istio-system
        - kube-system
       logLevel: debug
-      cniConfFileName: "istio-cni.conf"
```
Comment (xpivarc) [inline cluster-provision/gocli/opts/istio/manifests/istio-operator-with-cnao.cr.yaml:25]: Please just mention in commit that this reverts to previous configuration

**Review by xpivarc — COMMENTED** [review]
@aerosouund I have left few questions

File: cluster-up/cluster/k8s-provider-common.sh (hunk 6 above)
Comment (xpivarc) [reply cluster-up/cluster/k8s-provider-common.sh]: @aerosouund Let's minimize the changes so we fix the issues first and then we will improve

File: cluster-up/cluster/k8s-provider-common.sh (hunk 7 above)
Comment (aerosouund) [reply cluster-up/cluster/k8s-provider-common.sh:19]: it was there before

File: cluster-provision/gocli/cmd/run.go (hunk 4 above)
Comment (aerosouund) [reply cluster-provision/gocli/cmd/run.go:815]: indeed it can be enhanced, will do

**Comment by aerosouund** [comment]
@xpivarc 
kindly check if the commit messages now are expressive enough :)

**Comment by aerosouund** [comment]
@xpivarc 
i have added test cases to assert that the CNAO opt is handling the creation of multus properly
let me know if there are any further changes required

**Review by xpivarc — APPROVED** [review]

File: cluster-provision/gocli/opts/cnao/manifests/operator.yaml
Comment (oshoval) [inline cluster-provision/gocli/opts/cnao/manifests/operator.yaml:1]: iiuc now all the providers have the same file ? 
it means we cant have separated cnao anymore for each provider?
at the past the untangling of the folders was done deliberately for that
please consider if this is the desired behavior
if there will be problems, it would block us, we can either fix it or be proactive if it happens
thanks
